janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/JanusGraphPSerializer.java

```diff
@@ -0,0 +1,128 @@
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.apache.tinkerpop.gremlin.process.traversal.P;
+import org.apache.tinkerpop.gremlin.process.traversal.util.AndP;
+import org.apache.tinkerpop.gremlin.process.traversal.util.ConnectiveP;
+import org.apache.tinkerpop.gremlin.process.traversal.util.OrP;
+import org.apache.tinkerpop.shaded.kryo.Kryo;
+import org.apache.tinkerpop.shaded.kryo.Serializer;
+import org.apache.tinkerpop.shaded.kryo.io.Input;
+import org.apache.tinkerpop.shaded.kryo.io.Output;
+import org.janusgraph.core.attribute.Geo;
+import org.janusgraph.core.attribute.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Serialize Janus Graph Predicates.
+ * 
+ * Since all predicates end up being an instance of P and the TinkerPop P
+ * serializer is not extensible, it is copied here and the JanusGraph predicates
+ * are added.
+ *
+ */
+public class JanusGraphPSerializer extends Serializer<P> {
+
+    private static final Logger log = LoggerFactory.getLogger(JanusGraphPSerializer.class);
+
+    @Override
+    public void write(Kryo kryo, Output output, P p) {
+        output.writeString(
+                p instanceof ConnectiveP ? (p instanceof AndP ? "and" : "or") : p.getBiPredicate().toString());
+        if (p instanceof ConnectiveP || p.getValue() instanceof Collection) {
+            output.writeByte((byte) 0);
+            final Collection<?> coll = p instanceof ConnectiveP ? ((ConnectiveP<?>) p).getPredicates()
+                    : (Collection) p.getValue();
+            output.writeInt(coll.size());
+            coll.forEach(v -> kryo.writeClassAndObject(output, v));
+        } else {
+            output.writeByte((byte) 1);
+            kryo.writeClassAndObject(output, p.getValue());
+        }
+    }
+
+    @Override
+    public P read(Kryo kryo, Input input, Class<P> aClass) {
+        final String predicate = input.readString();
+        final boolean isCollection = input.readByte() == (byte) 0;
+        final Object value;
+        if (isCollection) {
+            value = new ArrayList();
+            final int size = input.readInt();
+            for (int ix = 0; ix < size; ix++) {
+                ((List) value).add(kryo.readClassAndObject(input));
+            }
+        } else {
+            value = kryo.readClassAndObject(input);
+        }
+
+        try {
+
+            if (value instanceof Collection) {
+                switch (predicate) {
+                case "and":
+                    return new AndP((List<P>) value);
```

Comment:
Should the "and" and "or" cases be in the else (not collection) case as in [PSerializer.read](https://github.com/apache/tinkerpop/blob/dfba83f046cc1c54ddc5d2ee26a8228c7af6bf1b/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/io/gryo/GryoSerializers.java#L209-L210)?